### ISSUE TITLE
bufferTimeout error on back pressure

### ISSUE THREAD

**Issue by vladgon** [reactor/reactor-core#2236]
## Expected Behavior
Should flush only when there is a request, skip flush otherwise 

## Actual Behavior
Errors when there is no downstream request, should skip flushing instead, similar issue probably with windowTimeout, didn't check though

## Steps to Reproduce
```java
@Test
public void testDelay() {
        final AtomicInteger count = new AtomicInteger(3);
        Flux.range(0, 1_000)
            .log("Range.", Level.FINE, SignalType.REQUEST)
            .doOnRequest(req -> log.debug("On Request {}", req))
            .log("", Level.FINE, SignalType.ON_NEXT)
            .bufferTimeout(1000, Duration.ofMillis(10))
            .doOnNext(w -> log.debug("Got buffer {}", w.size()))
            .flatMapSequential(integers -> Flux.fromIterable(integers)
                                               .flatMap(i -> Mono.delay(Duration.ofSeconds((int) (Math.random() * 30) / 10 + 1))
                                                                 .doFinally(signalType -> log.debug("Batch Complete {}", i)))
                                               .subscribeOn(Schedulers.parallel()),
                               3)
            .blockLast();
}
```

## Possible Solution
[Please see pull request](https://github.com/vladgon/reactor-core/pull/1/files)
## Your Environment
OS X, Windows 10

* Reactor version(s) used: reactor-core-3.3.10.RELEASE
* JVM version (`javar -version`): 1.8.0_211
* OS and version (eg `uname -a`):  OS X

**Comment by simonbasle** [comment]
timeout-based buffering (or windowing) is susceptible to backpressure overflow like that, because there is no notion of backpressure for a clock. you instruct the operator to forcefully emit at a given clock rate, so if the downstream cannot consume at this rate it gets terminated.

**Comment by vladgon** [comment]
In my opinion it should not terminate but rather skip if downstream is not ready to consume. The downstream doesn't and should not care about the source implementation e.i. buffer based, time-buffer based or anything else. Reactive stream spec  promise to respect backpressure - if there is no request then the source should block and wait. At least it should be an option for this behavior.

Thanks.

**Comment by vladgon** [comment]
The issue is still have not bee resolved. What should be a correct way to bufferTimeout infinite stream  in IO bound slow consumer e.g. DataBase. Let's consider a case of processing Kafka messages very fast rate and persisting to some slow storage let's say Elastic Search, RDMS

**Comment by chemicL** [comment]
The "fair backpressure" mechanism for `bufferTimeout` was implemented in #3332 and is available in 3.5.7.

**Comment by He-Pin** [comment]
![image](https://github.com/reactor/reactor-core/assets/501740/4980de44-1716-4470-ad8a-c4f5036abb9a)

This is a really bug, Cause a heavy production issue in my System, I posted a code in Pekko Stream.

I think if you want to advertise reactor-core as a Reactive library, you should implement it in the right way!


@vladgon cc

**Comment by chemicL** [comment]
Hey, @He-Pin. I'm sorry to hear about your production issue. Have you tried using [the new variant](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#bufferTimeout-int-java.time.Duration-boolean-) that provides backpressure against the timing events? Its signature is the following:

```
bufferTimeout(int maxSize, Duration maxTime, boolean fairBackpressure)
```

The last argument, when `true` switches to the implementation that honours backpressure. There is a known issue that can appear in high concurrent scenarios with it that needs to be addressed, but once that is done I'm considering replacing the old implementation with the new and removing the argument. I'd be keen to learn what your thoughts are on this.

Regarding Pekko Stream – I am not familiar with that project, can you elaborate how the screenshot and the linked issue relate to this one?

**Comment by He-Pin** [comment]
@chemicL 
Indeed, it appeared suddenly during a traffic spike, and because I'm consuming from a message queue/broker, the rebalancing mechanism led to the error propagating across multiple machines. As for this new approach, I see that it performs prefetching, but I prefer not to have prefetching. At the moment, I'm hesitant to use reactor-core in production due to this issue. The code mentioned above is about buffering data locally when the downstream is temporarily unprepared, then immediately providing that data to the downstream upon its next request, instead of reactor-core's current behavior of erroring out and canceling the entire stream. This is the correct implementation. The notion that a clock cannot provide backpressure is incorrect.

**Comment by He-Pin** [comment]
I just checked the fairBackpressure one, which the `flushABuffer` is protected with `WIP`, I think which is right now, and after nearly 10 minutes run, no bug regported.

**Comment by He-Pin** [comment]
About the new one, I saw the queue is being used to accumulate and polling into the buffer once `flush` is needed, I think that can be optimized and reduce the copy from queue to buffer.

**Comment by chemicL** [comment]
Thanks a lot for the analysis and suggestions @He-Pin 🥇 
Because we need to maintain API and behaviour compatibility, I think it would be wise to consider designing a new operator replacing `bufferTimeout` that could be introduced in 3.7 and both existing implementations (or just the non-backpressured one) could be deprecated.
I'm all in favour of providing a switch to skip prefetching if that is something you need. As noted, it should be achievable with the existing implementation.
For the remaining points regarding performance, I'd love to see JMH tests that prove it one way or another (both for the operator composition and for replacing the queue implementation). We'd need to have the existing implementation and the proposed implementation and see.
Please open a new issue summarizing these points and we can start brainstorming/experimenting. Are you considering a contribution here?

**Comment by He-Pin** [comment]
Yes, we better introduce a new operator, eg bufferWeightedWithin or something like that, will do tomorrow.
